ec: General cleanup

leoetlino opened the pull request:
This PR tries to improve readability and clarity of the ec code:

- Move all of the ec functions into the Common::ec namespace.

- Give the public functions better names and some usage information.

- Move all of the "elt" related functions into an "elt" class including all of the arithmetic operations, so that the logic becomes clearer and feels less like assembly.

  This also makes it much more obvious what the parameters are, instead of only using unsigned char* (which doesn't tell anything about what the pointer is used for or the size).

- Similarly, add a new "Point" class and move point functions there. Overload the arithmetic operators to make calculations easier to read.

The changeset is unfortunately a bit larger than I'd have liked it to be, but splitting it into smaller commits would have resulted in much more extra work.

[pywii has a variant of the original ec code](https://github.com/fail0verflow/hbc/blob/2fb61e432eed2fffb9f24a089067a7c2ec95fb1f/pywii/Common/pywii/ec.py#L50) that is easier to read, and it's what I used to check for mistakes during the cleanup.

I've also added unit tests for ec functions (to check for mistakes). All tests pass so there shouldn't be any 
 regression.

Warepire commented on Source/Core/Common/Crypto/bn.cpp:
```diff
@@ -8,21 +8,19 @@
 }
 
-int bn_compare(const u8* a, const u8* b, u32 n)
+int bn_compare(const u8* a, const u8* b, std::size_t n)
```
Can't this entire function be collapsed into a clamped `memcmp`?

leoetlino replied:
Done. (my original goal was to clean up ec, not bn :P)

Warepire commented on Source/Core/Common/Crypto/ec.cpp:
```diff
@@ -20,314 +21,213 @@
-  itoh_tsujii(s, t, t, 116);
-  elt_square(d, s);
+  Elt d;
+  for (size_t i = 0; i < std::tuple_size<decltype(Elt::data)>{}; i++)
```
`std::size_t`?

leoetlino commented on Source/Core/Common/Crypto/ec.cpp:
```diff
@@ -20,314 +21,213 @@
-static void itoh_tsujii(u8* d, const u8* a, const u8* b, u32 j)
-{
-  u8 t[30];
+  Elt ItohTsujii(const Elt& b, size_t j) const
```
For consistency with bn, I've changed ec to use std::size_t as well.

leoetlino commented on Source/Core/Common/Crypto/bn.h:
```diff
@@ -8,9 +8,9 @@
-void bn_mul(u8* d, const u8* a, const u8* b, const u8* N, u32 n);
-void bn_inv(u8* d, const u8* a, const u8* N, u32 n);  // only for prime N
-void bn_exp(u8* d, const u8* a, const u8* N, u32 n, const u8* e, u32 en);
+int bn_compare(const u8* a, const u8* b, std::size_t n);
```
You're right, it doesn't (I thought it would). `<cstddef>` is now included.

Tilka commented:
Would it be possible to use mbedtls for this?

leoetlino commented:
Unfortunately no. mbedtls does not support ECC-B233 (which is the curve used by IOS). We would have to replace mbedtls with openssl if we wanted to use a crypto library.

leoetlino commented:
Signing seems to be broken. Investigating.

leoetlino commented:
Fixed a regression that was introduced by the bn loop commit (silly off-by-one).

Also added yet another unit test to verify that signatures are valid.

akien-mga commented:
Just noting it down before looking further, the new `Common::ec` namespace introduced in 355b1b5 seems to cause a build issue on my system with GCC 5.5.0:
```
[  5%] Building CXX object Source/Core/Common/CMakeFiles/common.dir/Crypto/ec.cpp.o
cd /home/akien/Mageia/Checkout/dolphin-emu/BUILD/dolphin-5.0+7862/build/Source/Core/Common && /bin/ccache /bin/c++  -DDATA_DIR=\"/usr/share/dolphin-emu/\" -DHAVE_FFMPEG -DHAVE_LIBEVDEV=1 -$
In file included from /home/akien/Mageia/Checkout/dolphin-emu/BUILD/dolphin-5.0+7862/Source/Core/Common/Crypto/ec.cpp:17:0:
/home/akien/Mageia/Checkout/dolphin-emu/BUILD/dolphin-5.0+7862/Source/Core/Common/Crypto/ec.h:11:17: error: expected '{' before '::' token
 namespace Common::ec
                 ^
/home/akien/Mageia/Checkout/dolphin-emu/BUILD/dolphin-5.0+7862/Source/Core/Common/Crypto/ec.h:11:19: error: 'ec' in namespace '::' does not name a type
 namespace Common::ec
                   ^
/home/akien/Mageia/Checkout/dolphin-emu/BUILD/dolphin-5.0+7862/Source/Core/Common/Crypto/ec.cpp:24:17: error: expected '{' before '::' token
 namespace Common::ec
                 ^
/home/akien/Mageia/Checkout/dolphin-emu/BUILD/dolphin-5.0+7862/Source/Core/Common/Crypto/ec.cpp:24:19: error: 'ec' in namespace '::' does not name a type
 namespace Common::ec
                   ^
/home/akien/Mageia/Checkout/dolphin-emu/BUILD/dolphin-5.0+7862/Source/Core/Common/Crypto/ec.cpp:321:1: error: expected '}' at end of input
 }  // namespace Common::ec
 ^
/home/akien/Mageia/Checkout/dolphin-emu/BUILD/dolphin-5.0+7862/Source/Core/Common/Crypto/ec.cpp:321:1: error: expected '}' at end of input
Source/Core/Common/CMakeFiles/common.dir/build.make:281: recipe for target 'Source/Core/Common/CMakeFiles/common.dir/Crypto/ec.cpp.o' failed
```
It builds fine with GCC 7.3.0.

akien-mga commented:
Looks like nested namespace definitions are a [C++17 feature](http://www.nuonsoft.com/blog/2017/08/01/c17-nested-namespaces/) which is [only available in GCC 6.0 and later](https://gcc.gnu.org/projects/cxx-status.html#cxx17).

So you'd have to either:
- up the minimal GCC version required in https://github.com/dolphin-emu/dolphin/blob/master/CMakeLists.txt#L206
- or avoid using nested namespaces for now, using instead `namespace Common { namespace ec { } }`
